Improved output/workflow of 'verify' (including 'pull')

WestleyK opened the pull request:
## Description of the Pull Request (PR):

This PR improves the output of `verify` when verifying a container. Suggested by @gmkurtzer.

### Example:

```bash
$ singularity pull library://westleyk/demo/signed-test
INFO:    Downloading library image
 2.61 MiB / 2.61 MiB [===========================================================================================] 100.00% 4.23 MiB/s 0s
ERROR:   unable to verify container: signed-test_latest.sif
This image is not signed, and thus its contents cannot be verified.

$ singularity verify signed-test_latest.sif 
Container is signed by 6 key(s):

Verifying signature F: 8883491F4268F173C6E5DC49EDECE4F3F38D871E:
[LOCAL]    Sylabs Admin <support@sylabs.io>

Verifying signature F: 36E5D93B03B37178CDC9E9D7AA246EFBB99BBE05:
[MISSING]  key does not exist in local, or remote keystore

Verifying signature F: F69C21F759C8EA06FD32CCF4536523CE1E109AF3:
[LOCAL]    E2E test key (E2E test key; do not use) <westley@sylabs.io>

Verifying signature F: C7E7C8C3635DD06930669A2283B30190FEEF8162:
[LOCAL]    Westley (from gpg) <westley@sylabs.io>

Verifying signature F: E09D378BD534368E8BF2390386216BF0A30345A0:
[REMOTE]   WestleyK (test key 1) <westley@sylabs.io>

Verifying signature F: C7E7C8C3635DD06930669A2283B30190FEEF8162:
[LOCAL]    Westley (from gpg) <westley@sylabs.io>

FATAL:   Failed to verify: signed-test_latest.sif: unble to fetch key 36E5D93B03B37178CDC9E9D7AA246EFBB99BBE05: no matching keys found for fingerprint
```

And a more common alpine container:

```bash
$ singularity pull library://alpine
INFO:    Downloading library image
 2.59 MiB / 2.59 MiB [==============================================================] 100.00% 3.43 MiB/s 0s
INFO:    Container is trusted - run 'singularity key list' to list your trusted keys
INFO:    Download complete: alpine_latest.sif

$ singularity verify alpine_latest.sif 
Container is signed by 1 key(s):

Verifying signature F: 8883491F4268F173C6E5DC49EDECE4F3F38D871E:
[LOCAL]    Sylabs Admin <support@sylabs.io>

Signature(s) and data integrity verified
```

_And of course, there will also be some color in the example, but its not shown here._

## This fixes or addresses the following GitHub issues:

- Fixes #3646 

<br>

Commit:
Updated output

Commit:
Updated Missink key message

Commit:
Merge remote-tracking branch 'upstream/master' into verify_fix

Commit:
Cleaned up warnings

Commit:
Updated file

Commit:
Shortend error message (no newline) & print key ID

Commit:
Added [TRUSTED] to list of signed keys

Commit:
fixxing Travis

jchavezlavalle commented on pkg/signing/signing.go:
```diff
@@ -248,6 +248,7 @@ func IsSigned(cpath, keyServerURI string, id uint32, isGroup bool, authToken str
 	}
 	if noLocalKey {
 		return true, fmt.Errorf("no local key matching entity")
+		//return true, nil
```
remove this line

jchavezlavalle commented on pkg/signing/signing.go:
```diff
@@ -308,18 +311,23 @@ func Verify(cpath, keyServiceURI string, id uint32, isGroup bool, authToken stri
+			trusted = false
+			notLocalKey = true
+
 			// if theres a error, thats proboly becuse we dont have a local key
```
check typos on comments

jchavezlavalle commented on pkg/signing/signing.go:
```diff
@@ -280,6 +281,8 @@ func Verify(cpath, keyServiceURI string, id uint32, isGroup bool, authToken stri
 	// compare freshly computed hash with hashes stored in signatures block(s)
 	for _, v := range signatures {
+		trusted := true
+
```
remove additional lines

Commit:
Merge remote-tracking branch 'upstream/master' into verify_fix

Commit:
Updated output of verify

Commit:
Addressing some comments

Commit:
Remove the "exit 1 if container is signed, but no local key" (will not edit 1)

Commit:
Fixed CCI...

Commit:
Fixed pull_test.go -> no need to test exit 1 if pulled not default container

Commit:
Updated pull_test.go -> no need to download the public key, improved --dir flag testing by tmpDirReturn(t)

Commit:
Holy :hankey: travis finds a problome with everything! _looks at @WestleyK_ + updated table format - easier to read and debug

Commit:
Attempting to fix Travis...

Commit:
Merge remote-tracking branch 'upstream/master' into verify_fix

Commit:
Merge remote-tracking branch 'upstream/master' into verify_fix

Commit:
Updated comments + removed extra commented code + added ERROR when '-l' flag and theres no local key

Commit:
Improved output of 'verify' by adding color, more below
Improving the output of verify by adding some color to [TRUSTED] [REMOTE] [MISSING]. Also when pulling a container, this will now no longer show the ouput of the 'verify', it will just print if its "untrusted"/unsiged.

Commit:
Empty commit. Just kicking CCI

Commit:
Merge remote-tracking branch 'upstream/master' into verify_fix
And resolved some conflict...

Commit:
Added info when pulling a container if its "trusted"
also, removed '.' in aborting message

Commit:
Dont "exit 1" while verifying a container (if theres no local key)

Commit:
Fixing output, adding newline spaces to seperate the keys so its easier to read
These changes was suggested by @ikaneshiro. Also call 'sylog.Fatalf()' if theres no key in the keystore (missing key)

Commit:
Removed 'data integrity checked...' and added 'Container signed by X keys:'
Also suggested by IanK. + removed commented code

ikaneshiro commented on cmd/singularity/pull_test.go:
```diff
@@ -56,16 +50,20 @@ func imagePull(t *testing.T, library, pullDir string, imagePath string, sourceSp
 }
 
+// tmpDirReturn will return a tmp dir path in /tmp.
+func tmpDirReturn(t *testing.T) string {
```
Can we change this to `makeTmpDir` or something to that effect? Having `Return` in the name isn't very descriptive.

WestleyK replied:
Actually, this should not even be in this PR, I'll make a new PR that improves the `pull_test.go`

WestleyK replied:
Nevermined, these changes need to be in this PR (this PR changes the exit code).

ikaneshiro commented on go.mod:
```diff
@@ -34,7 +34,7 @@ require (
 	github.com/docker/libtrust v0.0.0-20160708172513-aabc10ec26b7 // indirect
 	github.com/dsnet/compress v0.0.1 // indirect
-	github.com/fatih/color v1.7.0 // indirect
+	github.com/fatih/color v1.7.1-0.20181010231311-3f9d52f7176a
```
Is there a reason you're using a specific commit of this package instead of using a tagged version?

WestleyK replied:
Umm, I guess this vendor was added after I made my PR (it was not there a couple weeks ago, or I probably did not see it)... I'll change it back.

ikaneshiro reviewed (changes requested):
Couple nit picks but I think this is well on its way. Also, can you add some sort of success message when container verification succeeds? like `INFO: Container signatures and data integrity verified`

Commit:
Reverted color vendor to 1.7.0 + go mod tidy

Commit:
Changed function name from tmpDirReturn => makeTmpDir

Commit:
Improved "quiet" handling + added
"Container signatures and data integrity verified" if the container is trusted (have all signed keys locally)

Commit:
Now appending the error messages for "failed downloading key"

Commit:
Merge remote-tracking branch 'upstream/master' into verify_fix

Commit:
Improving "trusted" and public keys docs

Commit:
Final improvments to return error

Commit:
Print the "Signature(s) and data integrity verified" after the keys that signed that container

Commit:
Changed up return error message like it was before (without appending the error)
This was causing a issue when verifying a container thats not signed, the error message would be hard to read. (fixed now)

Commit:
Changed "[TRUSTED]" to "[LOCAL]" & removed user prompt when pulling a unsigend container
As Greg suggested

Commit:
Small fix for printing "[LOCAL]"
Trying to force push this, not worth a extra commit

Commit:
Fixed comment for ErrLibraryPullUnsigned

Commit:
Resolving conflict in pull_test.go
Merge remote-tracking branch 'upstream/master' into verify_fix

Commit:
Check ErrUnknownIssuer before attempting to download the remote key
Me And IanK were talking about this, it should be checking this error before downloading a key from the keystore. I know theres probably going to be more refactoring in the 'Verify()' function, but wanted to push my local changes before the weekend.

Commit:
go mod vendor

Commit:
Improving the output for verify/pull

Commit:
Aligned output for verify

Commit:
Still improving the output...

Commit:
Still improving the outout...

Commit:
go mod tidy

Commit:
Improving output, again

Commit:
Removed unused commented code

Commit:
Merge remote-tracking branch 'upstream/master' into verify_fix
